### ISSUE TITLE
io.EOF should never be flagged, as it should never be wrapped

### ISSUE THREAD

**Issue by zikaeroh** [polyfloyd/go-errorlint#4]
`io.EOF` should never be wrapped; it is only correct to check it exactly in order to fulfill the API of the `io` interfaces. It'd be best to ignore any comparisons to `io.EOF` (or potentially flag the opposite).

This is made clearer in the new [Go 1.16 docs](https://tip.golang.org/pkg/io/#pkg-variables), see also the discussion here: https://github.com/golang/go/issues/40827#issuecomment-701503761

**Comment by coder543** [comment]
The situation is a bit more nuanced than that:

> I don't think the issue is that io.EOF shouldn't be wrapped ever, it's that it shouldn't be wrapped when implementing io.Reader

> In other words, the constraint here is not a property of io.EOF but a property of io.Reader so asking something purely about io.EOF can't give a useful answer.

Quoting [from here](https://github.com/golang/go/issues/40827#issuecomment-675341885)

If this special case is implemented, it should only be enforced inside an `io.Reader` implementation. It’s perfectly fine to wrap `io.EOF` elsewhere.

**Comment by zikaeroh** [comment]
You're referring to returning `io.EOF` by wrapping it, but I'm more concerned with code that uses functions/methods that return `io.EOF` and are flagged as wrong for making the comparison when they actually are correct.

It'd certainly be nicer to flag potential wraps of `io.EOF` in `Read` methods, for example, but as far as I'm aware, this library does not check anything on the wrapping side except the formatting string. The rest of the checks are checks at consumption time where it'd be provable that what's being compared is `io.EOF`.

See also: https://github.com/golang/go/issues/39155#issuecomment-701611876

**Comment by zikaeroh** [comment]
I'm only noting this now given golangci-lint added this analysis to their suite; I personally had a lot of false positives in my codebase due to known errors like `io.EOF` and `sql.ErrNoRows` and I've just disabled the linter outright. (Not that it doesn't have value to someone, of course!)

**Comment by coder543** [comment]
To be clear, I’m not the author of this lint, just someone who likes the way it currently works.

`errors.Is` has never once shown up on a flamegraph that I’ve captured, so the developer cost of inconsistency with regard to how errors are checked is much more worrisome to me than an imperceptible performance hit from always checking errors with `Is` and `As`.

`ErrNoRows` wasn’t mentioned in that discussion, so how do you know it is guaranteed never to be wrapped? I just read the documentation, and I think I see how you came to that conclusion, but I wonder if there’s a list of such errors that the linter could reference, or if this I would be a manual effort to sift through the stdlib documentation and hope no one ever violates this restriction when implementing one of these interfaces?

**Comment by zikaeroh** [comment]
The argument for not using `errors.Is` isn't so much that it's costly, but that using it might hide bugs that might otherwise lurk in the code. E.g., you may write code that works well, until you use some stdlib function that assumes the non-wrapping of an error. By checking explicitly for things like `io.EOF`, you'll fail faster rather than have some nebulous problem deep in someone else's code. The linked discussions talk about the different pros/cons likely better than I can write out myself.

I didn't offer `sql.ErrNoRows` as one that is guaranteed not to be wrapped, just one that I know to be a false positive in my codebase. I use SQLBoiler, which is guaranteed to return `sql.ErrNoRows` from its library and the generated code that gets put into my codebase. In that context, it too is an error I know I should be checking explicitly to conform with what the library says it does.

All-in-all, I just want to be able to write "correct" code without too much linter fuss. Clearly, I'm content with just disabling the linter, I only filed this for the specific case of `io.EOF` that had been recently decided/clarified as I knew it'd be good to report sooner rather than later once everyone's attempted to rewrite their code to appease `golangci-lint` (and potentially transformed it suboptimally).

**Comment by polyfloyd** [comment]
Heyo, author here :wave: 

Not wrapping `io.EOF`, `sql.ErrNoRows`, etc is indeed considered valid. I have been working on an allowlist to at least have this linter not complain about the standard library, but it's not finished yet!

My attempt can be found here: https://github.com/polyfloyd/go-errorlint/commit/f0e303dc81c6ef5c75f0e8b1527b73da76a24401

I hope I will be able to have some time to complete this, but that will be unlikely until at least 2021. If someone urgently needs this, they could give this a shot themselves of course :)

**Comment by polyfloyd** [comment]
Hey all,

I have just completed the patch I mentioned half a year ago and pushed it to master.

Errors that are documented to be returned from standard library functions will no longer trigger warnings by the linter. It will still complain about `io.EOF` returned from 3rd party functions though, but there is nothing from stopping them to still wrap it. When we introduce a lint that forbids wrapping `io.EOF` I feel confident enough to add a blanket implementation that allows comparing to `io.EOF` everywhere.

**Comment by kolyshkin** [comment]
I think this is addressed by https://github.com/polyfloyd/go-errorlint/commit/3e7f851e8f223a4d2ff3d97b5cc4b0686765e5ef, and so the issue can be closed.

**Comment by polyfloyd** [comment]
That is correct! I had kept it open to also include a patch that adds a check that disallows wrapping `io.EOF`, but the main culprit has been resolved and we can always open a new issue for that.

**Comment by apprehensions** [comment]
this still happens with `go-errorlint version devel comments-go-here buildID=78cf27ec8de86066bfaa660dd1bc0a1bde06e8918480a897d422a661fc452d17`
```go

	reader := tar.NewReader(stream)

	for {
		header, err := reader.Next()

		if err == io.EOF {
			break
		}
```

**Comment by polyfloyd** [comment]
@apprehensions Fixed, thanks for reporting

**Comment by kolyshkin** [comment]
Fixed by commit 06e615c687a5f9f2ade419f3da7f9cba338f9cbf; not yet released.

**Comment by polyfloyd** [comment]
@kolyshkin This fix was released in v1.4.1

**Comment by mitar** [comment]
I see that this is now implemented by checking where `io.EOF` is coming from. Can you please just provide an option or something to never flag `== io.EOF`? I am reading from `net/http.Response.Body` and comparing with `io.EOF` and I am still getting errors. I think you should just allow `== io.EOF` and this is it.
